torchtune/models/qwen2/_tokenizer.py

```diff
@@ -23,13 +23,34 @@
     "<|im_start|>": 151644,
     "<|im_end|>": 151645,
 }
-
+QWEN2_5_SPECIAL_TOKENS = {
```

Comment:
The tokenizer will handle the tokens if they're present in the data, but it won't insert them. Like for example, there's no way for us to reliably detect a tool call in the data and add `<tool_call>` special tokens around it, so we need the user to preprocess their dataset in a way that matches Qwen's preferred tool-calling format. Maybe we can add this information in the documentation? Other than the `<|im_*|>` tokens, the only special tokens we can proactively insert into the data, as far as I can tell, are the `vision_start/end` ones.

Regarding your second point, I changed `QwenTokenizer` back to `Qwen2Tokenizer`, but `QWEN2_SPECIAL_TOKENS` and `QWEN2_5_SPECIAL_TOKENS` needs to remain separate, otherwise Qwen2 models will start seeing Qwen2.5 tokens.

Reply:
We can detect a tool call in the data. By the time we get to the tokenizer we can assume the user has already formatted their data into Messages, in which case role=assistant and ipython=True would indicate a tool call.

Reply:
If `ipython==True`, can we assume that the entire message is the tool call? Or is it possible for the message to contain other text as well?

Reply:
yes, I would say it's safe to assume this. If a user includes content in a Message that's not a tool call but has `ipython=True`, I would consider that a user error. We should be very clear in documentation though to guide users.